js/src/main/scala/scalajson/ast/JValue.scala

```diff
@@ -77,22 +77,26 @@ object JNumber {
   def apply(value: Float): JValue = value match {
     case n if java.lang.Float.isNaN(n) => JNull
     case n if n.isInfinity => JNull
-    case _ => JNumber(value.toString)
+    case _ => new JNumber(value.toString)
   }
+
+  def fromString(value: String): Option[JNumber] =
+    if (value.matches(jNumberRegex))
+      Some(new JNumber(value))
+    else
+      None
+
+  def unapply(arg: JNumber): Option[String] = Some(arg.underlying)
```

Comment:
This could be improved via something like [this](https://github.com/sirthias/swave/blob/master/core/src/main/scala/swave/core/util/AnyRefExtractor.scala), i.e. it could be done in a way that avoids the allocation here.
Not saying that this should be done, just pointing out the opportunity.